File: casper/src/test/scala/coop/rchain/casper/blocks/proposer/BlockCreatorSpec.scala
package coop.rchain.casper.blocks.proposer

import cats.syntax.all._
import com.google.protobuf.ByteString
import coop.rchain.blockstorage.{BlockStore, KeyValueBlockStore}
import coop.rchain.blockstorage.dag.BlockDagKeyValueStorage
import coop.rchain.blockstorage.deploy.{DeployStorage, KeyValueDeployStorage}
import coop.rchain.casper.{CasperShardConf, CasperSnapshot, OnChainCasperState, ValidatorIdentity}
import coop.rchain.casper.helper.TestBlockDagRepresentation
import coop.rchain.casper.protocol.DeployData
import coop.rchain.casper.util.GenesisBuilder.defaultValidatorSks
import coop.rchain.casper.util.rholang.{Resources, RuntimeManager}
import coop.rchain.crypto.signatures.{Secp256k1, Signed}
import coop.rchain.metrics.{Metrics, Span}
import coop.rchain.metrics.Metrics.MetricsNOP
import coop.rchain.p2p.EffectsTestInstances.LogStub
import coop.rchain.rholang
import coop.rchain.shared.{Log, Time}
import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import org.scalatest.{FlatSpec, Matchers}

import scala.concurrent.duration._

/**
  * Unit tests for BlockCreator.
  *
  * Tests the deploy preparation and cleanup logic.
  */
class BlockCreatorSpec extends FlatSpec with Matchers {

  implicit private val logStub: Log[Task]     = new LogStub[Task]()
  implicit private val metrics: Metrics[Task] = new MetricsNOP[Task]()
  implicit private val span: Span[Task]       = Span.noop[Task]
  implicit private val timeEff: Time[Task] = new Time[Task] {
    override def currentMillis: Task[Long]                   = Task.delay(System.currentTimeMillis())
    override def nanoTime: Task[Long]                        = Task.delay(System.nanoTime())
    override def sleep(duration: FiniteDuration): Task[Unit] = Task.sleep(duration)
  }

  private val validatorSk       = defaultValidatorSks.head
  private val validatorIdentity = ValidatorIdentity(validatorSk)
  private val validatorId       = ByteString.copyFrom(validatorIdentity.publicKey.bytes)

  private val deployLifespan = 50

  private def createDeploy(
      validAfterBlockNumber: Long,
      expirationTimestamp: Option[Long] = None
  ): Signed[DeployData] = {
    val deployData = DeployData(
      term = s"new x in { x!($validAfterBlockNumber) }",
      timestamp = System.currentTimeMillis(),
      phloPrice = 1,
      phloLimit = 1000,
      validAfterBlockNumber = validAfterBlockNumber,
      shardId = "test-shard",
      expirationTimestamp = expirationTimestamp
    )
    Signed(deployData, Secp256k1, validatorSk)
  }

  private def createSnapshot(maxBlockNum: Long): CasperSnapshot[Task] = {
    val dag = TestBlockDagRepresentation[Task]()

    val onChainState = OnChainCasperState(
      shardConf = CasperShardConf(
        faultToleranceThreshold = 0f,
        shardName = "test-shard",
        parentShardId = "",
        finalizationRate = 0,
        maxNumberOfParents = 10,
        maxParentDepth = 0,
        synchronyConstraintThreshold = 0f,
        heightConstraintThreshold = 0L,
        deployLifespan = deployLifespan,
        casperVersion = 1L,
        configVersion = 1L,
        bondMinimum = 0L,
        bondMaximum = Long.MaxValue,
        epochLength = 0,
        quarantineLength = 0,
        minPhloPrice = 0L,
        enableMergeableChannelGC = false,
        mergeableChannelsGCDepthBuffer = 10,
        disableLateBlockFiltering = false,
        disableValidatorProgressCheck = false
      ),
      bondsMap = Map(validatorId -> 100L),
      activeValidators = Seq(validatorId)
    )

    CasperSnapshot[Task](
      dag = dag,
      lastFinalizedBlock = ByteString.EMPTY,
      lca = ByteString.EMPTY,
      tips = IndexedSeq.empty,
      parents = List.empty,
      justifications = Set.empty,
      invalidBlocks = Map.empty,
      deploysInScope = Set.empty,
      maxBlockNum = maxBlockNum,
      maxSeqNums = Map(validatorId -> 0),
      onChainState = onChainState
    )
  }

  "BlockCreator.create" should "remove block-expired deploys while keeping valid ones in storage" in {
    val test = rholang.Resources
      .mkTempDir[Task]("block-creator-test-")
      .evalMap(Resources.mkTestRNodeStoreManager[Task])
      .use { kvm =>
        for {
          blockStore            <- KeyValueBlockStore[Task](kvm)
          _                     <- BlockDagKeyValueStorage.create[Task](kvm)
          deployStorageInstance <- KeyValueDeployStorage[Task](kvm)
          runtimeManager        <- Resources.mkRuntimeManagerAt[Task](kvm)

          result <- {
            implicit val ds: DeployStorage[Task]  = deployStorageInstance
            implicit val bs: BlockStore[Task]     = blockStore
            implicit val rm: RuntimeManager[Task] = runtimeManager

            for {
              // With deployLifespan = 50 and currentBlock = 101 (maxBlockNum = 100),
              // earliestBlockNumber = 101 - 50 = 51
              //
              // Expired deploy: validAfterBlockNumber = 0 (<= 51, expired)
              // Valid deploy: validAfterBlockNumber = 60 (> 51 and < 101, valid)
              expiredDeploy <- Task.delay(createDeploy(validAfterBlockNumber = 0L))
              validDeploy   <- Task.delay(createDeploy(validAfterBlockNumber = 60L))

              // Add both deploys to storage
              _ <- ds.add(List(expiredDeploy, validDeploy))

              // Verify both deploys are in storage
              deploysBeforeCreate <- ds.readAll
              _                   = deploysBeforeCreate.size shouldBe 2

              snapshot = createSnapshot(maxBlockNum = 100L)

              // Call BlockCreator.create
              // The cleanup happens in prepareUserDeploys before block creation
              // Block creation may fail due to empty parents, but that's after cleanup
              _ <- BlockCreator.create(snapshot, validatorIdentity).attempt

              // Verify: expired deploy removed, valid deploy kept
              deploysAfterCreate <- ds.readAll
              _                  = deploysAfterCreate.size shouldBe 1
              _                  = deploysAfterCreate.head.sig shouldBe validDeploy.sig
            } yield ()
          }
        } yield result
      }

    test.runSyncUnsafe()
  }

  it should "remove both block-expired and time-expired deploys while keeping valid ones" in {
    val test = rholang.Resources
      .mkTempDir[Task]("block-creator-test-")
      .evalMap(Resources.mkTestRNodeStoreManager[Task])
      .use { kvm =>
        for {
          blockStore            <- KeyValueBlockStore[Task](kvm)
          _                     <- BlockDagKeyValueStorage.create[Task](kvm)
          deployStorageInstance <- KeyValueDeployStorage[Task](kvm)
          runtimeManager        <- Resources.mkRuntimeManagerAt[Task](kvm)

          result <- {
            implicit val ds: DeployStorage[Task]  = deployStorageInstance
            implicit val bs: BlockStore[Task]     = blockStore
            implicit val rm: RuntimeManager[Task] = runtimeManager

            for {
              pastTimestamp <- Task.delay(System.currentTimeMillis() - 60000L) // 1 minute ago

              // Block-expired deploy (validAfterBlockNumber = 0 is expired)
              blockExpiredDeploy <- Task.delay(createDeploy(validAfterBlockNumber = 0L))

              // Time-expired deploy (validAfterBlockNumber = 60 is valid, but expirationTimestamp is past)
              timeExpiredDeploy <- Task.delay(
                                    createDeploy(
                                      validAfterBlockNumber = 60L,
                                      expirationTimestamp = Some(pastTimestamp)
                                    )
                                  )

              // Valid deploy (validAfterBlockNumber = 60 is valid, no expiration timestamp)
              validDeploy <- Task.delay(createDeploy(validAfterBlockNumber = 60L))

              // Add all deploys to storage
              _ <- ds.add(List(blockExpiredDeploy, timeExpiredDeploy, validDeploy))

              // Verify all deploys are in storage
              deploysBeforeCreate <- ds.readAll
              _                   = deploysBeforeCreate.size shouldBe 3

              snapshot = createSnapshot(maxBlockNum = 100L)

              _ <- BlockCreator.create(snapshot, validatorIdentity).attempt

              // Verify: both expired deploys removed, valid deploy kept
              deploysAfterCreate <- ds.readAll
              _                  = deploysAfterCreate.size shouldBe 1
              _                  = deploysAfterCreate.head.sig shouldBe validDeploy.sig
            } yield ()
          }
        } yield result
      }

    test.runSyncUnsafe()
  }
}

File: casper/tests/blocks/block_creator_spec.rs
// See casper/src/test/scala/coop/rchain/casper/blocks/proposer/BlockCreatorSpec.scala
//
// Unit tests for BlockCreator.
// Tests the deploy preparation and cleanup logic.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use block_storage::rust::{
    deploy::key_value_deploy_storage::KeyValueDeployStorage,
    key_value_block_store::KeyValueBlockStore,
};
use casper::rust::{
    blocks::proposer::block_creator,
    casper::{CasperShardConf, CasperSnapshot, OnChainCasperState},
    util::rholang::runtime_manager::RuntimeManager,
    validator_identity::ValidatorIdentity,
};
use crypto::rust::{
    private_key::PrivateKey,
    signatures::{secp256k1::Secp256k1, signed::Signed},
};
use dashmap::{DashMap, DashSet};
use models::rust::casper::protocol::casper_message::DeployData;
use models::ByteString;
use prost::bytes::Bytes;
use rspace_plus_plus::rspace::shared::{
    in_mem_store_manager::InMemoryStoreManager, key_value_store_manager::KeyValueStoreManager,
};

use crate::util::genesis_builder::DEFAULT_VALIDATOR_SKS;
use crate::util::rholang::resources;

const DEPLOY_LIFESPAN: i64 = 50;

/// Creates a signed deploy with the given parameters
fn create_deploy(
    valid_after_block_number: i64,
    expiration_timestamp: Option<i64>,
    validator_sk: &PrivateKey,
) -> Signed<DeployData> {
    let deploy_data = DeployData {
        term: format!("new x in {{ x!({}) }}", valid_after_block_number),
        time_stamp: SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_millis() as i64)
            .unwrap_or(0),
        phlo_price: 1,
        phlo_limit: 1000,
        valid_after_block_number,
        shard_id: "test-shard".to_string(),
        expiration_timestamp,
    };

    Signed::create(deploy_data, Box::new(Secp256k1), validator_sk.clone())
        .expect("Failed to create signed deploy")
}

/// Creates a CasperSnapshot for testing with the given parameters.
/// Uses an in-memory DAG representation (matching Scala's TestBlockDagRepresentation).
fn create_snapshot(max_block_num: i64, validator_id: Bytes) -> CasperSnapshot {
    let shard_conf = CasperShardConf {
        fault_tolerance_threshold: 0.0,
        shard_name: "test-shard".to_string(),
        parent_shard_id: "".to_string(),
        finalization_rate: 0,
        max_number_of_parents: 10,
        max_parent_depth: 0,
        synchrony_constraint_threshold: 0.0,
        height_constraint_threshold: 0,
        deploy_lifespan: DEPLOY_LIFESPAN,
        casper_version: 1,
        config_version: 1,
        bond_minimum: 0,
        bond_maximum: i64::MAX,
        epoch_length: 0,
        quarantine_length: 0,
        min_phlo_price: 0,
        enable_mergeable_channel_gc: false,
        mergeable_channels_gc_depth_buffer: 10,
        disable_late_block_filtering: false,
        disable_validator_progress_check: false,
        ..CasperShardConf::new()
    };

    let mut bonds_map: HashMap<ByteString, i64> = HashMap::new();
    bonds_map.insert(validator_id.clone(), 100);

    // Set maxSeqNums like Scala does: Map(validatorId -> 0)
    let max_seq_nums: DashMap<ByteString, u64> = DashMap::new();
    max_seq_nums.insert(validator_id.clone(), 0);

    let on_chain_state = OnChainCasperState {
        shard_conf,
        bonds_map,
        active_validators: vec![validator_id],
    };

    // Use in-memory DAG representation (like Scala's TestBlockDagRepresentation)
    let dag = resources::new_key_value_dag_representation();

    CasperSnapshot {
        dag,
        last_finalized_block: Bytes::new(),
        lca: Bytes::new(),
        tips: vec![],
        parents: vec![],
        justifications: DashSet::new(),
        invalid_blocks: HashMap::new(),
        deploys_in_scope: Arc::new(DashSet::new()),
        rejected_in_scope: Arc::new(DashSet::new()),
        max_block_num,
        max_seq_nums,
        on_chain_state,
    }
}

/// Test: "remove block-expired deploys while keeping valid ones in storage"
///
/// With deployLifespan = 50 and currentBlock = 101 (maxBlockNum = 100),
/// earliestBlockNumber = 101 - 50 = 51
///
/// Expired deploy: validAfterBlockNumber = 0 (<= 51, expired)
/// Valid deploy: validAfterBlockNumber = 60 (> 51 and < 101, valid)
#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn should_remove_block_expired_deploys_while_keeping_valid_ones() {
    crate::init_logger();

    let validator_sk = DEFAULT_VALIDATOR_SKS[0].clone();
    let validator_identity = ValidatorIdentity::new(&validator_sk);
    let validator_id: Bytes = validator_identity.public_key.bytes.clone().into();

    // Create all stores from a single InMemoryStoreManager (like Scala's kvm pattern)
    let mut kvm = InMemoryStoreManager::new();

    let deploy_storage = Arc::new(Mutex::new(
        KeyValueDeployStorage::new(&mut kvm)
            .await
            .expect("Failed to create deploy storage"),
    ));
    let rejected_deploy_buffer = Arc::new(Mutex::new(
        block_storage::rust::deploy::key_value_rejected_deploy_buffer::KeyValueRejectedDeployBuffer::new(&mut kvm)
            .await
            .expect("Failed to create rejected deploy buffer"),
    ));

    let block_store = KeyValueBlockStore::create_from_kvm(&mut kvm)
        .await
        .expect("Failed to create block store");

    let rspace_store = kvm
        .r_space_stores()
        .await
        .expect("Failed to get rspace store");
    let mergeable_store = resources::mergeable_store_from_dyn(&mut kvm)
        .await
        .expect("Failed to create mergeable store");

    let (mut runtime_manager, _) = RuntimeManager::create_with_history(
        rspace_store,
        mergeable_store,
        std::sync::Arc::new(casper::rust::genesis::genesis::Genesis::default_mergeable_tags()),
        rholang::rust::interpreter::external_services::ExternalServices::noop(),
    );

    // Create deploys:
    // - Expired deploy: validAfterBlockNumber = 0 (<= 51, expired)
    // - Valid deploy: validAfterBlockNumber = 60 (> 51 and < 101, valid)
    let expired_deploy = create_deploy(0, None, &validator_sk);
    let valid_deploy = create_deploy(60, None, &validator_sk);

    // Add both deploys to storage
    {
        let mut ds = deploy_storage.lock().unwrap();
        ds.add(vec![expired_deploy.clone(), valid_deploy.clone()])
            .expect("Failed to add deploys");

        // Verify both deploys are in storage
        let deploys_before = ds.read_all().expect("Failed to read deploys");
        assert_eq!(deploys_before.len(), 2, "Expected 2 deploys before create");
    }

    // Create snapshot with maxBlockNum = 100
    let snapshot = create_snapshot(100, validator_id);

    // Call BlockCreator.create
    // The cleanup happens in prepareUserDeploys before block creation
    // Block creation may fail due to empty parents, but that's after cleanup
    let _ = block_creator::create(
        &snapshot,
        &validator_identity,
        None,
        deploy_storage.clone(),
        rejected_deploy_buffer.clone(),
        &mut runtime_manager,
        &mut block_store.clone(),
        false,
    )
    .await;

    // Verify: expired deploy removed, valid deploy kept
    {
        let ds = deploy_storage.lock().unwrap();
        let deploys_after = ds.read_all().expect("Failed to read deploys");
        assert_eq!(
            deploys_after.len(),
            1,
            "Expected 1 deploy after create (expired should be removed)"
        );

        let remaining_deploy = deploys_after.iter().next().unwrap();
        assert_eq!(
            remaining_deploy.sig, valid_deploy.sig,
            "Expected valid deploy to remain"
        );
    }
}

/// Test: "remove both block-expired and time-expired deploys while keeping valid ones"
///
/// - Block-expired deploy (validAfterBlockNumber = 0 is expired)
/// - Time-expired deploy (validAfterBlockNumber = 60 is valid, but expirationTimestamp is past)
/// - Valid deploy (validAfterBlockNumber = 60 is valid, no expiration timestamp)
#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn should_remove_both_block_expired_and_time_expired_deploys() {
    crate::init_logger();

    let validator_sk = DEFAULT_VALIDATOR_SKS[0].clone();
    let validator_identity = ValidatorIdentity::new(&validator_sk);
    let validator_id: Bytes = validator_identity.public_key.bytes.clone().into();

    // Create all stores from a single InMemoryStoreManager (like Scala's kvm pattern)
    let mut kvm = InMemoryStoreManager::new();

    let deploy_storage = Arc::new(Mutex::new(
        KeyValueDeployStorage::new(&mut kvm)
            .await
            .expect("Failed to create deploy storage"),
    ));
    let rejected_deploy_buffer = Arc::new(Mutex::new(
        block_storage::rust::deploy::key_value_rejected_deploy_buffer::KeyValueRejectedDeployBuffer::new(&mut kvm)
            .await
            .expect("Failed to create rejected deploy buffer"),
    ));

    let block_store = KeyValueBlockStore::create_from_kvm(&mut kvm)
        .await
        .expect("Failed to create block store");

    let rspace_store = kvm
        .r_space_stores()
        .await
        .expect("Failed to get rspace store");
    let mergeable_store = resources::mergeable_store_from_dyn(&mut kvm)
        .await
        .expect("Failed to create mergeable store");

    let (mut runtime_manager, _) = RuntimeManager::create_with_history(
        rspace_store,
        mergeable_store,
        std::sync::Arc::new(casper::rust::genesis::genesis::Genesis::default_mergeable_tags()),
        rholang::rust::interpreter::external_services::ExternalServices::noop(),
    );

    // 1 minute ago (past timestamp for time-expired deploy)
    let past_timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
        - 60000;

    // Create deploys:
    // - Block-expired deploy (validAfterBlockNumber = 0 is expired)
    // - Time-expired deploy (validAfterBlockNumber = 60 is valid, but expirationTimestamp is past)
    // - Valid deploy (validAfterBlockNumber = 60 is valid, no expiration timestamp)
    let block_expired_deploy = create_deploy(0, None, &validator_sk);
    let time_expired_deploy = create_deploy(60, Some(past_timestamp), &validator_sk);
    let valid_deploy = create_deploy(60, None, &validator_sk);

    // Add all deploys to storage
    {
        let mut ds = deploy_storage.lock().unwrap();
        ds.add(vec![
            block_expired_deploy.clone(),
            time_expired_deploy.clone(),
            valid_deploy.clone(),
        ])
        .expect("Failed to add deploys");

        // Verify all deploys are in storage
        let deploys_before = ds.read_all().expect("Failed to read deploys");
        assert_eq!(deploys_before.len(), 3, "Expected 3 deploys before create");
    }

    // Create snapshot with maxBlockNum = 100
    let snapshot = create_snapshot(100, validator_id);

    // Call BlockCreator.create
    let _ = block_creator::create(
        &snapshot,
        &validator_identity,
        None,
        deploy_storage.clone(),
        rejected_deploy_buffer.clone(),
        &mut runtime_manager,
        &mut block_store.clone(),
        false,
    )
    .await;

    // Verify: both expired deploys removed, valid deploy kept
    {
        let ds = deploy_storage.lock().unwrap();
        let deploys_after = ds.read_all().expect("Failed to read deploys");
        assert_eq!(
            deploys_after.len(),
            1,
            "Expected 1 deploy after create (both expired should be removed)"
        );

        let remaining_deploy = deploys_after.iter().next().unwrap();
        assert_eq!(
            remaining_deploy.sig, valid_deploy.sig,
            "Expected valid deploy to remain"
        );
    }
}

/// Test: expired sigs are purged from the rejected-deploy buffer too.
///
/// Sets up a deploy that is in the rejected-deploy buffer but NOT in
/// deploy_storage — the realistic scenario after a sig has been
/// conflict-rejected by the merge engine and the original storage entry
/// has aged out via prior expired-removal sweeps. Without the buffer
/// purge, a sustained-load adversary that keeps generating conflicts
/// can grow the buffer's LMDB store unbounded — the read path filters
/// expired sigs out so they aren't re-proposed, but on-disk entries
/// would persist.
///
/// snapshot maxBlockNum = 100, deployLifespan = 50 → earliestBlockNumber = 51
/// Expired buffer deploy: validAfterBlockNumber = 0 (<= 51, expired)
/// Valid buffer deploy:   validAfterBlockNumber = 60 (> 51, valid)
#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn should_remove_expired_deploys_from_rejected_deploy_buffer() {
    crate::init_logger();

    let validator_sk = DEFAULT_VALIDATOR_SKS[0].clone();
    let validator_identity = ValidatorIdentity::new(&validator_sk);
    let validator_id: Bytes = validator_identity.public_key.bytes.clone().into();

    let mut kvm = InMemoryStoreManager::new();

    let deploy_storage = Arc::new(Mutex::new(
        KeyValueDeployStorage::new(&mut kvm)
            .await
            .expect("Failed to create deploy storage"),
    ));
    let rejected_deploy_buffer = Arc::new(Mutex::new(
        block_storage::rust::deploy::key_value_rejected_deploy_buffer::KeyValueRejectedDeployBuffer::new(&mut kvm)
            .await
            .expect("Failed to create rejected deploy buffer"),
    ));

    let block_store = KeyValueBlockStore::create_from_kvm(&mut kvm)
        .await
        .expect("Failed to create block store");

    let rspace_store = kvm
        .r_space_stores()
        .await
        .expect("Failed to get rspace store");
    let mergeable_store = resources::mergeable_store_from_dyn(&mut kvm)
        .await
        .expect("Failed to create mergeable store");

    let (mut runtime_manager, _) = RuntimeManager::create_with_history(
        rspace_store,
        mergeable_store,
        std::sync::Arc::new(casper::rust::genesis::genesis::Genesis::default_mergeable_tags()),
        rholang::rust::interpreter::external_services::ExternalServices::noop(),
    );

    let expired_deploy = create_deploy(0, None, &validator_sk);
    let valid_deploy = create_deploy(60, None, &validator_sk);

    // Buffer-only: deploys live in the rejected-deploy buffer, not in
    // deploy_storage. This is the regression-relevant case — without
    // the buffer-purge fix, the expired sig persists in LMDB.
    {
        let mut buf = rejected_deploy_buffer.lock().unwrap();
        buf.add(vec![expired_deploy.clone(), valid_deploy.clone()])
            .expect("Failed to add deploys to buffer");

        let deploys_before = buf.read_all().expect("Failed to read buffer");
        assert_eq!(
            deploys_before.len(),
            2,
            "Expected 2 deploys in buffer before create"
        );
    }

    let snapshot = create_snapshot(100, validator_id);

    let _ = block_creator::create(
        &snapshot,
        &validator_identity,
        None,
        deploy_storage.clone(),
        rejected_deploy_buffer.clone(),
        &mut runtime_manager,
        &mut block_store.clone(),
        false,
    )
    .await;

    // Verify: expired sig removed from buffer, valid sig retained.
    {
        let buf = rejected_deploy_buffer.lock().unwrap();
        assert!(
            !buf.contains_sig(&expired_deploy.sig)
                .expect("Failed to query buffer for expired sig"),
            "Expired sig must NOT remain in the rejected-deploy buffer after create. \
             If this fires, the expired-removal sweep in `prepare_user_deploys` is not \
             extending to the buffer — sustained-load adversaries can grow the buffer \
             unbounded."
        );
        assert!(
            buf.contains_sig(&valid_deploy.sig)
                .expect("Failed to query buffer for valid sig"),
            "Valid (unexpired) sig must remain in the rejected-deploy buffer"
        );
    }
}
